source/server/listener_manager_impl.cc

```diff
@@ -89,19 +91,49 @@ ListenerImpl::ListenerImpl(const envoy::api::v2::Listener& config, ListenerManag
       local_drain_manager_(parent.factory_.createDrainManager(config.drain_type())) {
   // TODO(htuch): Support multiple filter chains #1280, add constraint to ensure we have at least on
   // filter chain #1308.
-  ASSERT(config.filter_chains().size() == 1);
-  const auto& filter_chain = config.filter_chains()[0];
-
-  listener_scope_ =
-      parent_.server_.stats().createScope(fmt::format("listener.{}.", address_->asString()));
-
-  if (filter_chain.has_tls_context()) {
-    Ssl::ServerContextConfigImpl context_config(filter_chain.tls_context());
-    ssl_context_ = parent_.server_.sslContextManager().createSslServerContext(*listener_scope_,
-                                                                              context_config);
+  ASSERT(config.filter_chains().size() >= 1);
+
+  // Skip lookup and update of the SSL Context if there is only one filter chain
+  // and it doesn't enforce any SNI restrictions.
+  const bool skip_context_update =
+      (config.filter_chains().size() == 1 &&
+       config.filter_chains()[0].filter_chain_match().sni_domains().empty());
+
+  size_t filters_hash = 0;
+  uint32_t has_tls = 0;
+  uint32_t has_stk = 0;
+  for (const auto& filter_chain : config.filter_chains()) {
+    std::vector<std::string> sni_domains(filter_chain.filter_chain_match().sni_domains().begin(),
+                                         filter_chain.filter_chain_match().sni_domains().end());
+    if (filters_hash == 0) {
+      filters_hash = RepeatedPtrUtil::hash(filter_chain.filters());
+      filter_factories_ = parent_.factory_.createFilterFactoryList(filter_chain.filters(), *this);
+    } else if (filters_hash != RepeatedPtrUtil::hash(filter_chain.filters())) {
+      throw EnvoyException(fmt::format("error adding listener '{}': use of different filter chains "
```

Comment:
nit: I would say "use of SNI and different filter chains..."

Reply:
Technically, existing code allows multiple filter chains, as long as they have the same filters, regardless of use of SNI, so I think that existing text is a bit more correct.

Reply:
OK that's fine.